{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating a collection data \n",
    "\n",
    "In this notebook, the main focuses is collecting and wrangling all the data for our series of processes. The data gathered are from two main databases known as [DBatVir](http://www.mgc.ac.cn/DBatVir/) and [NCBI Genome](https://www.ncbi.nlm.nih.gov/genome/viruses/). \n",
    "\n",
    "- DBatVIR -> database that contains known viral families associated with each bat\n",
    "- NCBI -> used for obtaining genome and gene information\n",
    "\n",
    "##  step 1: Collect \n",
    "The first step is to gather all the data from the databases and save it to the ./db folder. The `./db` will serve as our database when ever need to access data. This will prevent constantly sending requests to the databases. \n",
    "\n",
    "\n",
    "## Step 2: Filter\n",
    "\n",
    "## Step 3: Profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "from typing import Union\n",
    "from pathlib import Path\n",
    "import gzip\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "# vseek imports\n",
    "sys.path.append(\"../\")\n",
    "import vseek.common.vseek_paths as vsp\n",
    "from vseek.apis.dbat_vir_db import collect_dbatvir_data\n",
    "from vseek.apis.ncbi import get_all_viral_accessions, get_viral_genomes\n",
    "from vseek.common.loader import load_bat_virus_data, load_geolocations, load_human_ppi, load_species_atlas\n",
    "from vseek.apis.ncbi import get_viral_genes, get_taxa_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ADD INPUTS HERE \n",
    "email = \"erikishere3@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Viruses</th>\n",
       "      <th>Viral family</th>\n",
       "      <th>From Bat</th>\n",
       "      <th>Bat diet type</th>\n",
       "      <th>Bat family</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Collection year</th>\n",
       "      <th>Sampling country</th>\n",
       "      <th>Sequence coding</th>\n",
       "      <th>References</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bat adeno-associated virus 07YN</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>unclassified Chiroptera</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>Unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bat adeno-associated virus 09YN</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>unclassified Chiroptera</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2009</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>Unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bat adeno-associated virus 1003-HB-Mr</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>Myotis ricketti</td>\n",
       "      <td>Insectivore and piscivore</td>\n",
       "      <td>Vespertilionidae</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>J Gen Virol 2010, 91(Pt 10):2601-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bat adeno-associated virus 1008-HB-Mr</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>Myotis ricketti</td>\n",
       "      <td>Insectivore and piscivore</td>\n",
       "      <td>Vespertilionidae</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>J Gen Virol 2010, 91(Pt 10):2601-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bat adeno-associated virus 1019-HB-Rs</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>Rhinolophus sinicus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>J Gen Virol 2010, 91(Pt 10):2601-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13957</th>\n",
       "      <td>Rhinolophus pusillus coronavirus ZSR42</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13958</th>\n",
       "      <td>Rhinolophus pusillus norovirus ZSR43</td>\n",
       "      <td>Caliciviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13959</th>\n",
       "      <td>Rhinolophus pusillus norovirus ZSR45</td>\n",
       "      <td>Caliciviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13960</th>\n",
       "      <td>Rhinolophus pusillus coronavirus ZSR6</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13961</th>\n",
       "      <td>Table saved from ZOVER [Fri Jun 10 2022]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13962 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Viruses   Viral family  \\\n",
       "0               Bat adeno-associated virus 07YN   Parvoviridae   \n",
       "1               Bat adeno-associated virus 09YN   Parvoviridae   \n",
       "2         Bat adeno-associated virus 1003-HB-Mr   Parvoviridae   \n",
       "3         Bat adeno-associated virus 1008-HB-Mr   Parvoviridae   \n",
       "4         Bat adeno-associated virus 1019-HB-Rs   Parvoviridae   \n",
       "...                                         ...            ...   \n",
       "13957    Rhinolophus pusillus coronavirus ZSR42  Coronaviridae   \n",
       "13958      Rhinolophus pusillus norovirus ZSR43  Caliciviridae   \n",
       "13959      Rhinolophus pusillus norovirus ZSR45  Caliciviridae   \n",
       "13960     Rhinolophus pusillus coronavirus ZSR6  Coronaviridae   \n",
       "13961  Table saved from ZOVER [Fri Jun 10 2022]            NaN   \n",
       "\n",
       "                      From Bat              Bat diet type        Bat family  \\\n",
       "0      unclassified Chiroptera                        NaN               NaN   \n",
       "1      unclassified Chiroptera                        NaN               NaN   \n",
       "2              Myotis ricketti  Insectivore and piscivore  Vespertilionidae   \n",
       "3              Myotis ricketti  Insectivore and piscivore  Vespertilionidae   \n",
       "4          Rhinolophus sinicus                Insectivore     Rhinolophidae   \n",
       "...                        ...                        ...               ...   \n",
       "13957     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13958     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13959     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13960     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13961                      NaN                        NaN               NaN   \n",
       "\n",
       "      Sample type Collection year Sampling country Sequence coding  \\\n",
       "0           Feces            2007            China             Cap   \n",
       "1           Feces            2009            China             Cap   \n",
       "2           Feces            2007            China             Cap   \n",
       "3           Feces            2007            China             Cap   \n",
       "4           Feces            2007            China             Cap   \n",
       "...           ...             ...              ...             ...   \n",
       "13957         Mix            2015            China            RdRp   \n",
       "13958         Mix            2015            China            RdRp   \n",
       "13959         Mix            2015            China            RdRp   \n",
       "13960         Mix            2015            China            RdRp   \n",
       "13961         NaN             NaN              NaN             NaN   \n",
       "\n",
       "                               References  \n",
       "0                             Unpublished  \n",
       "1                             Unpublished  \n",
       "2      J Gen Virol 2010, 91(Pt 10):2601-9  \n",
       "3      J Gen Virol 2010, 91(Pt 10):2601-9  \n",
       "4      J Gen Virol 2010, 91(Pt 10):2601-9  \n",
       "...                                   ...  \n",
       "13957            Sci Rep 2017, 7(1):10917  \n",
       "13958            Sci Rep 2017, 7(1):10917  \n",
       "13959            Sci Rep 2017, 7(1):10917  \n",
       "13960            Sci Rep 2017, 7(1):10917  \n",
       "13961                                 NaN  \n",
       "\n",
       "[13962 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: the data has been downloaded already\n",
    "# collect the data from the VBatDB\n",
    "batvir_df = collect_dbatvir_data()\n",
    "batvir_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Representative</th>\n",
       "      <th>Neighbor</th>\n",
       "      <th>Host</th>\n",
       "      <th>Selected lineage</th>\n",
       "      <th>Taxonomy name</th>\n",
       "      <th>Segment name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>HQ420896</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>KY463519</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>HQ420897</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>MK035759</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>KY569019</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259836</th>\n",
       "      <td>NC_062761</td>\n",
       "      <td>MZ334528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Halorubrum virus HRTV-28</td>\n",
       "      <td>Halorubrum virus HRTV-28</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259837</th>\n",
       "      <td>NC_062762</td>\n",
       "      <td>MZ334526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Halorubrum virus HRTV-29</td>\n",
       "      <td>Halorubrum virus HRTV-29</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259838</th>\n",
       "      <td>NC_062763</td>\n",
       "      <td>MZ334501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae,Haloferacalesvirus,Halorubrum virus...</td>\n",
       "      <td>Halorubrum virus HSTV-4</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259839</th>\n",
       "      <td>NC_060136</td>\n",
       "      <td>OK040786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Siphoviridae,,Gordonia phage Kudefre</td>\n",
       "      <td>Gordonia phage Kudefre</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259840</th>\n",
       "      <td>NC_054649</td>\n",
       "      <td>MT768060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Swiduovirus,Swiduovirus swi2</td>\n",
       "      <td>Escherichia phage vB_EcoS_swi2</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259841 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Representative  Neighbor               Host  \\\n",
       "0           NC_003663  HQ420896  human,vertebrates   \n",
       "1           NC_003663  KY463519  human,vertebrates   \n",
       "2           NC_003663  HQ420897  human,vertebrates   \n",
       "3           NC_003663  MK035759  human,vertebrates   \n",
       "4           NC_003663  KY569019  human,vertebrates   \n",
       "...               ...       ...                ...   \n",
       "259836      NC_062761  MZ334528                NaN   \n",
       "259837      NC_062762  MZ334526                NaN   \n",
       "259838      NC_062763  MZ334501                NaN   \n",
       "259839      NC_060136  OK040786                NaN   \n",
       "259840      NC_054649  MT768060                NaN   \n",
       "\n",
       "                                         Selected lineage  \\\n",
       "0                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "1                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "2                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "3                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "4                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "...                                                   ...   \n",
       "259836                           Halorubrum virus HRTV-28   \n",
       "259837                           Halorubrum virus HRTV-29   \n",
       "259838  Myoviridae,Haloferacalesvirus,Halorubrum virus...   \n",
       "259839               Siphoviridae,,Gordonia phage Kudefre   \n",
       "259840                       Swiduovirus,Swiduovirus swi2   \n",
       "\n",
       "                         Taxonomy name Segment name  \n",
       "0                         Cowpox virus      segment  \n",
       "1                         Cowpox virus      segment  \n",
       "2                         Cowpox virus      segment  \n",
       "3                         Cowpox virus      segment  \n",
       "4                         Cowpox virus      segment  \n",
       "...                                ...          ...  \n",
       "259836        Halorubrum virus HRTV-28      segment  \n",
       "259837        Halorubrum virus HRTV-29      segment  \n",
       "259838         Halorubrum virus HSTV-4      segment  \n",
       "259839          Gordonia phage Kudefre      segment  \n",
       "259840  Escherichia phage vB_EcoS_swi2      segment  \n",
       "\n",
       "[259841 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: the data has been downloaded already\n",
    "# downloading ncbi accession list\n",
    "ncbi_acc_df = get_all_viral_accessions()\n",
    "ncbi_acc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fomatting functions \n",
    "def split_lineages(lineage: str) -> list:\n",
    "    split_lineage = lineage.split(\",\")\n",
    "    family, genus = split_lineage[0], split_lineage[1]\n",
    "    if family == \"\":\n",
    "        return pd.Series([np.nan, genus])\n",
    "    elif genus == \"\":\n",
    "        return pd.Series([family, np.nan])\n",
    "    elif family == \"\" and genus == \"\":\n",
    "        return pd.Series([np.nan, np.nan])\n",
    "    return pd.Series([family, genus])\n",
    "\n",
    "\n",
    "def clean_accession(accession: str) -> str:\n",
    "    if accession is np.nan:\n",
    "        return np.nan\n",
    "\n",
    "    accession = accession.split(\",\")\n",
    "    if len(accession) > 0:\n",
    "        return accession[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered_bat_virus already exists. Skipping\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Representative</th>\n",
       "      <th>Host</th>\n",
       "      <th>Taxonomy name</th>\n",
       "      <th>Segment name</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001348</td>\n",
       "      <td>human</td>\n",
       "      <td>Human alphaherpesvirus 3</td>\n",
       "      <td>segment</td>\n",
       "      <td>Herpesviridae</td>\n",
       "      <td>Varicellovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001352</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 2</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Alphapapillomavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001354</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 41</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Nupapillomavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001356</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 1a</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Mupapillomavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001430</td>\n",
       "      <td>human</td>\n",
       "      <td>Enterovirus D68</td>\n",
       "      <td>segment</td>\n",
       "      <td>Picornaviridae</td>\n",
       "      <td>Enterovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NC_055340</td>\n",
       "      <td>human</td>\n",
       "      <td>Echarate virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NC_055341</td>\n",
       "      <td>human</td>\n",
       "      <td>Echarate virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NC_055342</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>NC_055343</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>NC_055344</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Representative   Host                 Taxonomy name Segment name  \\\n",
       "0        NC_001348  human      Human alphaherpesvirus 3      segment   \n",
       "1        NC_001352  human   Human papillomavirus type 2      segment   \n",
       "2        NC_001354  human  Human papillomavirus type 41      segment   \n",
       "3        NC_001356  human  Human papillomavirus type 1a      segment   \n",
       "4        NC_001430  human               Enterovirus D68      segment   \n",
       "..             ...    ...                           ...          ...   \n",
       "218      NC_055340  human                Echarate virus      segment   \n",
       "219      NC_055341  human                Echarate virus      segment   \n",
       "220      NC_055342  human               Maldonado virus      segment   \n",
       "221      NC_055343  human               Maldonado virus      segment   \n",
       "222      NC_055344  human               Maldonado virus      segment   \n",
       "\n",
       "               family                genus  \n",
       "0       Herpesviridae       Varicellovirus  \n",
       "1    Papillomaviridae  Alphapapillomavirus  \n",
       "2    Papillomaviridae     Nupapillomavirus  \n",
       "3    Papillomaviridae     Mupapillomavirus  \n",
       "4      Picornaviridae          Enterovirus  \n",
       "..                ...                  ...  \n",
       "218     Phenuiviridae          Phlebovirus  \n",
       "219     Phenuiviridae          Phlebovirus  \n",
       "220     Phenuiviridae          Phlebovirus  \n",
       "221     Phenuiviridae          Phlebovirus  \n",
       "222     Phenuiviridae          Phlebovirus  \n",
       "\n",
       "[223 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#save paths\n",
    "save_path = Path(vsp.db_path()) / \"filtered_bat_virus.csv.gz\"\n",
    "\n",
    "# filter viral family that are only found in bats\n",
    "viral_fam = batvir_df[\"Viral family\"].dropna().unique().tolist()\n",
    "\n",
    "\n",
    "if not save_path.is_file():\n",
    "    # filtering main data frame to only viruses found in all bats \n",
    "    sel_dfs = []\n",
    "    for v_fam in viral_fam:\n",
    "        sel_df = ncbi_acc_df.loc[ncbi_acc_df[\"Selected lineage\"].str.contains(v_fam)]\n",
    "        if len(sel_df) == 0:\n",
    "            continue\n",
    "        sel_dfs.append(sel_df)\n",
    "\n",
    "\n",
    "    # selecting viral genomes found in bats that are also found in human\n",
    "    sel_dfs = pd.concat(sel_dfs, axis=0)\n",
    "    sel_dfs = sel_dfs.loc[sel_dfs[\"Host\"].notnull()]\n",
    "    sel_dfs = sel_dfs.loc[sel_dfs[\"Host\"].isin([\"human\"])]\n",
    "\n",
    "    # splitting the selected lineage into Family, Genius \n",
    "    sel_dfs[[\"family\", \"genus\"]] = sel_dfs[\"Selected lineage\"].apply(split_lineages)\n",
    "    sel_dfs[\"Representative\"] = sel_dfs[\"Representative\"].apply(clean_accession)\n",
    "    sel_dfs = sel_dfs.drop([\"Selected lineage\"], axis=\"columns\")\n",
    "\n",
    "    # removing enties that do not have either a family or a genus entry\n",
    "    sel_dfs.loc[(sel_dfs[\"family\"].notnull() & (sel_dfs[\"genus\"].notnull()))]\n",
    "\n",
    "    # grouping data frames based on accession and removing duplicates\n",
    "    groups = sel_dfs.groupby(\"Representative\")\n",
    "    cleaned_dfs = []\n",
    "    for name, df in groups:\n",
    "        df = df.drop(\"Neighbor\", axis=\"columns\")\n",
    "        df = df.drop_duplicates(subset=[\"Representative\"])\n",
    "        cleaned_dfs.append(df)\n",
    "\n",
    "    # generating bat-human viral genome csv\n",
    "    init_db = vsp.init_db_path()\n",
    "    cleaned_dfs = pd.concat(cleaned_dfs, axis=0)\n",
    "    cleaned_dfs.to_csv(save_path, compression=\"gzip\", index=False)\n",
    "else:\n",
    "    print(\"filtered_bat_virus already exists. Skipping\") \n",
    "    cleaned_dfs = pd.read_csv(save_path)\n",
    "cleaned_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxon id csv exists already. Skipping ...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Representative</th>\n",
       "      <th>taxon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001348</td>\n",
       "      <td>10335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001352</td>\n",
       "      <td>337043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001354</td>\n",
       "      <td>10589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001356</td>\n",
       "      <td>334203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001430</td>\n",
       "      <td>138951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NC_055340</td>\n",
       "      <td>1000646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NC_055341</td>\n",
       "      <td>1000646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NC_055342</td>\n",
       "      <td>1004889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>NC_055343</td>\n",
       "      <td>1004889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>NC_055344</td>\n",
       "      <td>1004889</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Representative  taxon_id\n",
       "0        NC_001348     10335\n",
       "1        NC_001352    337043\n",
       "2        NC_001354     10589\n",
       "3        NC_001356    334203\n",
       "4        NC_001430    138951\n",
       "..             ...       ...\n",
       "218      NC_055340   1000646\n",
       "219      NC_055341   1000646\n",
       "220      NC_055342   1004889\n",
       "221      NC_055343   1004889\n",
       "222      NC_055344   1004889\n",
       "\n",
       "[223 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxon_save_path = Path(vsp.db_path()) / \"taxon_id.csv\"\n",
    "if not taxon_save_path.is_file():\n",
    "    # generating a taxon id profile \n",
    "    # NOTE: This will take a while to complete because there is a 1 second pause after each request\n",
    "    print(\"collecting taxon ids\")\n",
    "    c_accessions = cleaned_dfs[\"Representative\"].tolist()\n",
    "    collected_taxon_ids = []\n",
    "    for c_acc in c_accessions:\n",
    "        taxa_id = get_taxa_id(email=email, accession=c_acc, buffer=1.0)\n",
    "        result = [c_acc, taxa_id]\n",
    "        collected_taxon_ids.append(result)\n",
    "    df = pd.DataFrame(data=collected_taxon_ids, columns=[\"Representative\", \"taxon_id\"])\n",
    "    df.to_csv(taxon_save_path, index=False)\n",
    "else:\n",
    "    print(\"Taxon id csv exists already. Skipping ...\")\n",
    "    taxon_df = pd.read_csv(taxon_save_path)\n",
    "    taxon_df.columns = [\"Representative\", \"taxon_id\"]\n",
    "taxon_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final_filtered_bat_virus already exists. Skipping ...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Representative</th>\n",
       "      <th>Host</th>\n",
       "      <th>Taxonomy name</th>\n",
       "      <th>Segment name</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001348</td>\n",
       "      <td>human</td>\n",
       "      <td>Human alphaherpesvirus 3</td>\n",
       "      <td>segment</td>\n",
       "      <td>Herpesviridae</td>\n",
       "      <td>Varicellovirus</td>\n",
       "      <td>10335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001352</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 2</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Alphapapillomavirus</td>\n",
       "      <td>337043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001354</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 41</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Nupapillomavirus</td>\n",
       "      <td>10589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001356</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 1a</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Mupapillomavirus</td>\n",
       "      <td>334203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001430</td>\n",
       "      <td>human</td>\n",
       "      <td>Enterovirus D68</td>\n",
       "      <td>segment</td>\n",
       "      <td>Picornaviridae</td>\n",
       "      <td>Enterovirus</td>\n",
       "      <td>138951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NC_055340</td>\n",
       "      <td>human</td>\n",
       "      <td>Echarate virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "      <td>1000646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NC_055341</td>\n",
       "      <td>human</td>\n",
       "      <td>Echarate virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "      <td>1000646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NC_055342</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "      <td>1004889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>NC_055343</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "      <td>1004889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>NC_055344</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "      <td>1004889</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Representative   Host                 Taxonomy name Segment name  \\\n",
       "0        NC_001348  human      Human alphaherpesvirus 3      segment   \n",
       "1        NC_001352  human   Human papillomavirus type 2      segment   \n",
       "2        NC_001354  human  Human papillomavirus type 41      segment   \n",
       "3        NC_001356  human  Human papillomavirus type 1a      segment   \n",
       "4        NC_001430  human               Enterovirus D68      segment   \n",
       "..             ...    ...                           ...          ...   \n",
       "218      NC_055340  human                Echarate virus      segment   \n",
       "219      NC_055341  human                Echarate virus      segment   \n",
       "220      NC_055342  human               Maldonado virus      segment   \n",
       "221      NC_055343  human               Maldonado virus      segment   \n",
       "222      NC_055344  human               Maldonado virus      segment   \n",
       "\n",
       "               family                genus  taxon_id  \n",
       "0       Herpesviridae       Varicellovirus     10335  \n",
       "1    Papillomaviridae  Alphapapillomavirus    337043  \n",
       "2    Papillomaviridae     Nupapillomavirus     10589  \n",
       "3    Papillomaviridae     Mupapillomavirus    334203  \n",
       "4      Picornaviridae          Enterovirus    138951  \n",
       "..                ...                  ...       ...  \n",
       "218     Phenuiviridae          Phlebovirus   1000646  \n",
       "219     Phenuiviridae          Phlebovirus   1000646  \n",
       "220     Phenuiviridae          Phlebovirus   1004889  \n",
       "221     Phenuiviridae          Phlebovirus   1004889  \n",
       "222     Phenuiviridae          Phlebovirus   1004889  \n",
       "\n",
       "[223 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_cleaned_df_path = Path(vsp.db_path()) / \"final_filtered_bat_virus.csv.gz\"\n",
    "if not final_cleaned_df_path.is_file():\n",
    "    # merge cleaned_df and taxon\n",
    "    final_clean_df = cleaned_dfs.merge(taxon_df, on=\"Representative\")\n",
    "    final_clean_df.to_csv(final_cleaned_df_path, index=False)\n",
    "else:\n",
    "    print(\"final_filtered_bat_virus already exists. Skipping ...\")\n",
    "    final_clean_df = pd.read_csv(final_cleaned_df_path)\n",
    "final_clean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Building genome database ...\n",
      "Checking if genome database exists ...\n",
      "Genome database already exists. Checking for missing files\n",
      "No missing genomes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/Users/erikserrano/Development/prelim/prelim3/VSeek/db/genome'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accession_list = cleaned_dfs[\"Representative\"].tolist()[:5]\n",
    "get_viral_genomes(email = email, accessions=accession_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Downloading viral genes metadata\n",
      "Checking if genome database exists ...\n",
      "Database exists! Checking for missing gene data files\n"
     ]
    }
   ],
   "source": [
    "# collectiing viral gene meta data \n",
    "# loading bat-virus data\n",
    "bat_vir_df = load_bat_virus_data()\n",
    "all_accessions = bat_vir_df[\"Representative\"].tolist()\n",
    "\n",
    "bat_vir_df.head()\n",
    "\n",
    "get_viral_genes(email=email, accession=all_accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>country</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD</td>\n",
       "      <td>42.546245</td>\n",
       "      <td>1.601554</td>\n",
       "      <td>Andorra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AE</td>\n",
       "      <td>23.424076</td>\n",
       "      <td>53.847818</td>\n",
       "      <td>United Arab Emirates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF</td>\n",
       "      <td>33.939110</td>\n",
       "      <td>67.709953</td>\n",
       "      <td>Afghanistan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG</td>\n",
       "      <td>17.060816</td>\n",
       "      <td>-61.796428</td>\n",
       "      <td>Antigua and Barbuda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AI</td>\n",
       "      <td>18.220554</td>\n",
       "      <td>-63.068615</td>\n",
       "      <td>Anguilla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>YE</td>\n",
       "      <td>15.552727</td>\n",
       "      <td>48.516388</td>\n",
       "      <td>Yemen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>YT</td>\n",
       "      <td>-12.827500</td>\n",
       "      <td>45.166244</td>\n",
       "      <td>Mayotte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>ZA</td>\n",
       "      <td>-30.559482</td>\n",
       "      <td>22.937506</td>\n",
       "      <td>South Africa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>ZM</td>\n",
       "      <td>-13.133897</td>\n",
       "      <td>27.849332</td>\n",
       "      <td>Zambia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>ZW</td>\n",
       "      <td>-19.015438</td>\n",
       "      <td>29.154857</td>\n",
       "      <td>Zimbabwe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    country   latitude  longitude                  name\n",
       "0        AD  42.546245   1.601554               Andorra\n",
       "1        AE  23.424076  53.847818  United Arab Emirates\n",
       "2        AF  33.939110  67.709953           Afghanistan\n",
       "3        AG  17.060816 -61.796428   Antigua and Barbuda\n",
       "4        AI  18.220554 -63.068615              Anguilla\n",
       "..      ...        ...        ...                   ...\n",
       "240      YE  15.552727  48.516388                 Yemen\n",
       "241      YT -12.827500  45.166244               Mayotte\n",
       "242      ZA -30.559482  22.937506          South Africa\n",
       "243      ZM -13.133897  27.849332                Zambia\n",
       "244      ZW -19.015438  29.154857              Zimbabwe\n",
       "\n",
       "[245 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_geolocations()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Protein interactions data not found. Downloading ...\n",
      "\n",
      "Downloading all human protein interactions with all species\n",
      "Constructing ppi DataFrame\n",
      "Filtering to only viral-human protein interactions\n",
      "Annotating viral information\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_1</th>\n",
       "      <th>protein_1</th>\n",
       "      <th>species_2</th>\n",
       "      <th>protein_2</th>\n",
       "      <th>score</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>10335</td>\n",
       "      <td>NP04_VZVD</td>\n",
       "      <td>437</td>\n",
       "      <td>Varicella-zoster virus (strain Dumas)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>10255</td>\n",
       "      <td>F11_VAR67</td>\n",
       "      <td>188</td>\n",
       "      <td>Variola virus (isolate Human/India/Ind3/1967)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>10335</td>\n",
       "      <td>GM_VZVD</td>\n",
       "      <td>336</td>\n",
       "      <td>Varicella-zoster virus (strain Dumas)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>32603</td>\n",
       "      <td>GM_HHV6U</td>\n",
       "      <td>336</td>\n",
       "      <td>Human herpesvirus 6A (strain Uganda-1102)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>10359</td>\n",
       "      <td>GM_HCMVM</td>\n",
       "      <td>336</td>\n",
       "      <td>Human cytomegalovirus (strain Merlin)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11431391</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000473166</td>\n",
       "      <td>10335</td>\n",
       "      <td>GB_VZVD</td>\n",
       "      <td>227</td>\n",
       "      <td>Varicella-zoster virus (strain Dumas)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11432750</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000473172</td>\n",
       "      <td>10359</td>\n",
       "      <td>UL27_HCMVM</td>\n",
       "      <td>435</td>\n",
       "      <td>Human cytomegalovirus (strain Merlin)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11435910</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000473243</td>\n",
       "      <td>32603</td>\n",
       "      <td>GB_HHV6U</td>\n",
       "      <td>227</td>\n",
       "      <td>Human herpesvirus 6A (strain Uganda-1102)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11435913</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000473243</td>\n",
       "      <td>10335</td>\n",
       "      <td>GB_VZVD</td>\n",
       "      <td>227</td>\n",
       "      <td>Varicella-zoster virus (strain Dumas)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11435914</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSP00000473243</td>\n",
       "      <td>10359</td>\n",
       "      <td>GB_HCMVM</td>\n",
       "      <td>227</td>\n",
       "      <td>Human cytomegalovirus (strain Merlin)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19053 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         species_1        protein_1 species_2   protein_2 score  \\\n",
       "251           9606  ENSP00000000233     10335   NP04_VZVD   437   \n",
       "256           9606  ENSP00000000233     10255   F11_VAR67   188   \n",
       "261           9606  ENSP00000000233     10335     GM_VZVD   336   \n",
       "268           9606  ENSP00000000233     32603    GM_HHV6U   336   \n",
       "269           9606  ENSP00000000233     10359    GM_HCMVM   336   \n",
       "...            ...              ...       ...         ...   ...   \n",
       "11431391      9606  ENSP00000473166     10335     GB_VZVD   227   \n",
       "11432750      9606  ENSP00000473172     10359  UL27_HCMVM   435   \n",
       "11435910      9606  ENSP00000473243     32603    GB_HHV6U   227   \n",
       "11435913      9606  ENSP00000473243     10335     GB_VZVD   227   \n",
       "11435914      9606  ENSP00000473243     10359    GB_HCMVM   227   \n",
       "\n",
       "                                             annotation  \n",
       "251               Varicella-zoster virus (strain Dumas)  \n",
       "256       Variola virus (isolate Human/India/Ind3/1967)  \n",
       "261               Varicella-zoster virus (strain Dumas)  \n",
       "268           Human herpesvirus 6A (strain Uganda-1102)  \n",
       "269               Human cytomegalovirus (strain Merlin)  \n",
       "...                                                 ...  \n",
       "11431391          Varicella-zoster virus (strain Dumas)  \n",
       "11432750          Human cytomegalovirus (strain Merlin)  \n",
       "11435910      Human herpesvirus 6A (strain Uganda-1102)  \n",
       "11435913          Varicella-zoster virus (strain Dumas)  \n",
       "11435914          Human cytomegalovirus (strain Merlin)  \n",
       "\n",
       "[19053 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = load_human_ppi()\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: species atlas not found. Downlaoding...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>string_type</th>\n",
       "      <th>string_name_compact</th>\n",
       "      <th>official_name_ncbi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3702</td>\n",
       "      <td>core</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3711</td>\n",
       "      <td>core</td>\n",
       "      <td>Brassica rapa</td>\n",
       "      <td>Brassica rapa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3847</td>\n",
       "      <td>core</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>Glycine max</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4081</td>\n",
       "      <td>core</td>\n",
       "      <td>Solanum lycopersicum</td>\n",
       "      <td>Solanum lycopersicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4113</td>\n",
       "      <td>core</td>\n",
       "      <td>Solanum tuberosum</td>\n",
       "      <td>Solanum tuberosum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>1335626</td>\n",
       "      <td>core</td>\n",
       "      <td>Human coronavirus EMC (isolate United Kingdom/...</td>\n",
       "      <td>Human coronavirus EMC (isolate United Kingdom/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>1511779</td>\n",
       "      <td>core</td>\n",
       "      <td>Mischivirus A (isolate Miniopterus schreibersi...</td>\n",
       "      <td>Mischivirus A (isolate Miniopterus schreibersi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>1511900</td>\n",
       "      <td>core</td>\n",
       "      <td>Human parvovirus B19 (strain HV)</td>\n",
       "      <td>Human parvovirus B19 (strain HV)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>1511908</td>\n",
       "      <td>core</td>\n",
       "      <td>Murine minute virus (strain MVM prototype)</td>\n",
       "      <td>Murine minute virus (strain MVM prototype)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>1513235</td>\n",
       "      <td>core</td>\n",
       "      <td>Cafeteria roenbergensis virus (strain BV-PW1)</td>\n",
       "      <td>Cafeteria roenbergensis virus (strain BV-PW1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>232 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    taxon_id string_type                                string_name_compact  \\\n",
       "1       3702        core                               Arabidopsis thaliana   \n",
       "2       3711        core                                      Brassica rapa   \n",
       "3       3847        core                                        Glycine max   \n",
       "4       4081        core                               Solanum lycopersicum   \n",
       "5       4113        core                                  Solanum tuberosum   \n",
       "..       ...         ...                                                ...   \n",
       "228  1335626        core  Human coronavirus EMC (isolate United Kingdom/...   \n",
       "229  1511779        core  Mischivirus A (isolate Miniopterus schreibersi...   \n",
       "230  1511900        core                   Human parvovirus B19 (strain HV)   \n",
       "231  1511908        core         Murine minute virus (strain MVM prototype)   \n",
       "232  1513235        core      Cafeteria roenbergensis virus (strain BV-PW1)   \n",
       "\n",
       "                                    official_name_ncbi  \n",
       "1                                 Arabidopsis thaliana  \n",
       "2                                        Brassica rapa  \n",
       "3                                          Glycine max  \n",
       "4                                 Solanum lycopersicum  \n",
       "5                                    Solanum tuberosum  \n",
       "..                                                 ...  \n",
       "228  Human coronavirus EMC (isolate United Kingdom/...  \n",
       "229  Mischivirus A (isolate Miniopterus schreibersi...  \n",
       "230                   Human parvovirus B19 (strain HV)  \n",
       "231         Murine minute virus (strain MVM prototype)  \n",
       "232      Cafeteria roenbergensis virus (strain BV-PW1)  \n",
       "\n",
       "[232 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_atlas = load_species_atlas()\n",
    "species_atlas"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 ('VSeek')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "70cfca0adae6be6172ca3156d4c2b8dd3fe10450420e228861d3bfe418975f4b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
