{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating a collection data \n",
    "\n",
    "In this notebook, the main focuses is collecting and wrangling all the data for our series of processes. The data gathered are from two main databases known as [DBatVir](http://www.mgc.ac.cn/DBatVir/) and [NCBI Genome](https://www.ncbi.nlm.nih.gov/genome/viruses/). \n",
    "\n",
    "- DBatVIR -> database that contains known viral families associated with each bat\n",
    "- NCBI -> used for obtaining genome and gene information\n",
    "\n",
    "##  step 1: Collect \n",
    "The first step is to gather all the data from the databases and save it to the ./db folder. The `./db` will serve as our database when ever need to access data. This will prevent constantly sending requests to the databases. \n",
    "\n",
    "\n",
    "## Step 2: Filter\n",
    "\n",
    "## Step 3: Profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import glob\n",
    "from time import sleep\n",
    "from typing import Union\n",
    "from io import StringIO\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import entrezpy.conduit\n",
    "\n",
    "# vseek imports\n",
    "sys.path.append(\"../\")\n",
    "import vseek.common.vseek_paths as vsp\n",
    "from vseek.apis.dbat_vir_db import collect_dbatvir_data\n",
    "from vseek.apis.ncbi import get_all_viral_accessions\n",
    "from vseek.common.checks import genome_db_exist\n",
    "from vseek.common.io_files import genome_db_files, save_genome\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Viruses</th>\n",
       "      <th>Viral family</th>\n",
       "      <th>From Bat</th>\n",
       "      <th>Bat diet type</th>\n",
       "      <th>Bat family</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Collection year</th>\n",
       "      <th>Sampling country</th>\n",
       "      <th>Sequence coding</th>\n",
       "      <th>References</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bat adeno-associated virus 07YN</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>unclassified Chiroptera</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>Unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bat adeno-associated virus 09YN</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>unclassified Chiroptera</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2009</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>Unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bat adeno-associated virus 1003-HB-Mr</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>Myotis ricketti</td>\n",
       "      <td>Insectivore and piscivore</td>\n",
       "      <td>Vespertilionidae</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>J Gen Virol 2010, 91(Pt 10):2601-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bat adeno-associated virus 1008-HB-Mr</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>Myotis ricketti</td>\n",
       "      <td>Insectivore and piscivore</td>\n",
       "      <td>Vespertilionidae</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>J Gen Virol 2010, 91(Pt 10):2601-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bat adeno-associated virus 1019-HB-Rs</td>\n",
       "      <td>Parvoviridae</td>\n",
       "      <td>Rhinolophus sinicus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Feces</td>\n",
       "      <td>2007</td>\n",
       "      <td>China</td>\n",
       "      <td>Cap</td>\n",
       "      <td>J Gen Virol 2010, 91(Pt 10):2601-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13957</th>\n",
       "      <td>Rhinolophus pusillus coronavirus ZSR42</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13958</th>\n",
       "      <td>Rhinolophus pusillus norovirus ZSR43</td>\n",
       "      <td>Caliciviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13959</th>\n",
       "      <td>Rhinolophus pusillus norovirus ZSR45</td>\n",
       "      <td>Caliciviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13960</th>\n",
       "      <td>Rhinolophus pusillus coronavirus ZSR6</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>Rhinolophus pusillus</td>\n",
       "      <td>Insectivore</td>\n",
       "      <td>Rhinolophidae</td>\n",
       "      <td>Mix</td>\n",
       "      <td>2015</td>\n",
       "      <td>China</td>\n",
       "      <td>RdRp</td>\n",
       "      <td>Sci Rep 2017, 7(1):10917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13961</th>\n",
       "      <td>Table saved from ZOVER [Fri Jun 10 2022]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13962 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Viruses   Viral family  \\\n",
       "0               Bat adeno-associated virus 07YN   Parvoviridae   \n",
       "1               Bat adeno-associated virus 09YN   Parvoviridae   \n",
       "2         Bat adeno-associated virus 1003-HB-Mr   Parvoviridae   \n",
       "3         Bat adeno-associated virus 1008-HB-Mr   Parvoviridae   \n",
       "4         Bat adeno-associated virus 1019-HB-Rs   Parvoviridae   \n",
       "...                                         ...            ...   \n",
       "13957    Rhinolophus pusillus coronavirus ZSR42  Coronaviridae   \n",
       "13958      Rhinolophus pusillus norovirus ZSR43  Caliciviridae   \n",
       "13959      Rhinolophus pusillus norovirus ZSR45  Caliciviridae   \n",
       "13960     Rhinolophus pusillus coronavirus ZSR6  Coronaviridae   \n",
       "13961  Table saved from ZOVER [Fri Jun 10 2022]            NaN   \n",
       "\n",
       "                      From Bat              Bat diet type        Bat family  \\\n",
       "0      unclassified Chiroptera                        NaN               NaN   \n",
       "1      unclassified Chiroptera                        NaN               NaN   \n",
       "2              Myotis ricketti  Insectivore and piscivore  Vespertilionidae   \n",
       "3              Myotis ricketti  Insectivore and piscivore  Vespertilionidae   \n",
       "4          Rhinolophus sinicus                Insectivore     Rhinolophidae   \n",
       "...                        ...                        ...               ...   \n",
       "13957     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13958     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13959     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13960     Rhinolophus pusillus                Insectivore     Rhinolophidae   \n",
       "13961                      NaN                        NaN               NaN   \n",
       "\n",
       "      Sample type Collection year Sampling country Sequence coding  \\\n",
       "0           Feces            2007            China             Cap   \n",
       "1           Feces            2009            China             Cap   \n",
       "2           Feces            2007            China             Cap   \n",
       "3           Feces            2007            China             Cap   \n",
       "4           Feces            2007            China             Cap   \n",
       "...           ...             ...              ...             ...   \n",
       "13957         Mix            2015            China            RdRp   \n",
       "13958         Mix            2015            China            RdRp   \n",
       "13959         Mix            2015            China            RdRp   \n",
       "13960         Mix            2015            China            RdRp   \n",
       "13961         NaN             NaN              NaN             NaN   \n",
       "\n",
       "                               References  \n",
       "0                             Unpublished  \n",
       "1                             Unpublished  \n",
       "2      J Gen Virol 2010, 91(Pt 10):2601-9  \n",
       "3      J Gen Virol 2010, 91(Pt 10):2601-9  \n",
       "4      J Gen Virol 2010, 91(Pt 10):2601-9  \n",
       "...                                   ...  \n",
       "13957            Sci Rep 2017, 7(1):10917  \n",
       "13958            Sci Rep 2017, 7(1):10917  \n",
       "13959            Sci Rep 2017, 7(1):10917  \n",
       "13960            Sci Rep 2017, 7(1):10917  \n",
       "13961                                 NaN  \n",
       "\n",
       "[13962 rows x 10 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: the data has been downloaded already\n",
    "# collect the data from the VBatDB\n",
    "batvir_df = collect_dbatvir_data()\n",
    "batvir_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Representative</th>\n",
       "      <th>Neighbor</th>\n",
       "      <th>Host</th>\n",
       "      <th>Selected lineage</th>\n",
       "      <th>Taxonomy name</th>\n",
       "      <th>Segment name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>HQ420896</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>KY463519</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>HQ420897</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>MK035759</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_003663</td>\n",
       "      <td>KY569019</td>\n",
       "      <td>human,vertebrates</td>\n",
       "      <td>Poxviridae,Orthopoxvirus,Cowpox virus</td>\n",
       "      <td>Cowpox virus</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259836</th>\n",
       "      <td>NC_062761</td>\n",
       "      <td>MZ334528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Halorubrum virus HRTV-28</td>\n",
       "      <td>Halorubrum virus HRTV-28</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259837</th>\n",
       "      <td>NC_062762</td>\n",
       "      <td>MZ334526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Halorubrum virus HRTV-29</td>\n",
       "      <td>Halorubrum virus HRTV-29</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259838</th>\n",
       "      <td>NC_062763</td>\n",
       "      <td>MZ334501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae,Haloferacalesvirus,Halorubrum virus...</td>\n",
       "      <td>Halorubrum virus HSTV-4</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259839</th>\n",
       "      <td>NC_060136</td>\n",
       "      <td>OK040786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Siphoviridae,,Gordonia phage Kudefre</td>\n",
       "      <td>Gordonia phage Kudefre</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259840</th>\n",
       "      <td>NC_054649</td>\n",
       "      <td>MT768060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Swiduovirus,Swiduovirus swi2</td>\n",
       "      <td>Escherichia phage vB_EcoS_swi2</td>\n",
       "      <td>segment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259841 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Representative  Neighbor               Host  \\\n",
       "0           NC_003663  HQ420896  human,vertebrates   \n",
       "1           NC_003663  KY463519  human,vertebrates   \n",
       "2           NC_003663  HQ420897  human,vertebrates   \n",
       "3           NC_003663  MK035759  human,vertebrates   \n",
       "4           NC_003663  KY569019  human,vertebrates   \n",
       "...               ...       ...                ...   \n",
       "259836      NC_062761  MZ334528                NaN   \n",
       "259837      NC_062762  MZ334526                NaN   \n",
       "259838      NC_062763  MZ334501                NaN   \n",
       "259839      NC_060136  OK040786                NaN   \n",
       "259840      NC_054649  MT768060                NaN   \n",
       "\n",
       "                                         Selected lineage  \\\n",
       "0                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "1                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "2                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "3                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "4                   Poxviridae,Orthopoxvirus,Cowpox virus   \n",
       "...                                                   ...   \n",
       "259836                           Halorubrum virus HRTV-28   \n",
       "259837                           Halorubrum virus HRTV-29   \n",
       "259838  Myoviridae,Haloferacalesvirus,Halorubrum virus...   \n",
       "259839               Siphoviridae,,Gordonia phage Kudefre   \n",
       "259840                       Swiduovirus,Swiduovirus swi2   \n",
       "\n",
       "                         Taxonomy name Segment name  \n",
       "0                         Cowpox virus      segment  \n",
       "1                         Cowpox virus      segment  \n",
       "2                         Cowpox virus      segment  \n",
       "3                         Cowpox virus      segment  \n",
       "4                         Cowpox virus      segment  \n",
       "...                                ...          ...  \n",
       "259836        Halorubrum virus HRTV-28      segment  \n",
       "259837        Halorubrum virus HRTV-29      segment  \n",
       "259838         Halorubrum virus HSTV-4      segment  \n",
       "259839          Gordonia phage Kudefre      segment  \n",
       "259840  Escherichia phage vB_EcoS_swi2      segment  \n",
       "\n",
       "[259841 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: the data has been downloaded already\n",
    "# downloading ncbi accession list\n",
    "ncbi_acc_df = get_all_viral_accessions()\n",
    "ncbi_acc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fomatting functions \n",
    "def split_lineages(lineage: str) -> list:\n",
    "    split_lineage = lineage.split(\",\")\n",
    "    family, genus = split_lineage[0], split_lineage[1]\n",
    "    if family == \"\":\n",
    "        return pd.Series([np.nan, genus])\n",
    "    elif genus == \"\":\n",
    "        return pd.Series([family, np.nan])\n",
    "    elif family == \"\" and genus == \"\":\n",
    "        return pd.Series([np.nan, np.nan])\n",
    "    return pd.Series([family, genus])\n",
    "\n",
    "\n",
    "def clean_accession(accession: str) -> str:\n",
    "    if accession is np.nan:\n",
    "        return np.nan\n",
    "\n",
    "    accession = accession.split(\",\")\n",
    "    if len(accession) > 0:\n",
    "        return accession[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Representative</th>\n",
       "      <th>Host</th>\n",
       "      <th>Taxonomy name</th>\n",
       "      <th>Segment name</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>NC_001348</td>\n",
       "      <td>human</td>\n",
       "      <td>Human alphaherpesvirus 3</td>\n",
       "      <td>segment</td>\n",
       "      <td>Herpesviridae</td>\n",
       "      <td>Varicellovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214039</th>\n",
       "      <td>NC_001352</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 2</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Alphapapillomavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208580</th>\n",
       "      <td>NC_001354</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 41</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Nupapillomavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208577</th>\n",
       "      <td>NC_001356</td>\n",
       "      <td>human</td>\n",
       "      <td>Human papillomavirus type 1a</td>\n",
       "      <td>segment</td>\n",
       "      <td>Papillomaviridae</td>\n",
       "      <td>Mupapillomavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194084</th>\n",
       "      <td>NC_001430</td>\n",
       "      <td>human</td>\n",
       "      <td>Enterovirus D68</td>\n",
       "      <td>segment</td>\n",
       "      <td>Picornaviridae</td>\n",
       "      <td>Enterovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257436</th>\n",
       "      <td>NC_055340</td>\n",
       "      <td>human</td>\n",
       "      <td>Echarate virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257437</th>\n",
       "      <td>NC_055341</td>\n",
       "      <td>human</td>\n",
       "      <td>Echarate virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257464</th>\n",
       "      <td>NC_055342</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257465</th>\n",
       "      <td>NC_055343</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257463</th>\n",
       "      <td>NC_055344</td>\n",
       "      <td>human</td>\n",
       "      <td>Maldonado virus</td>\n",
       "      <td>segment</td>\n",
       "      <td>Phenuiviridae</td>\n",
       "      <td>Phlebovirus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Representative   Host                 Taxonomy name Segment name  \\\n",
       "810         NC_001348  human      Human alphaherpesvirus 3      segment   \n",
       "214039      NC_001352  human   Human papillomavirus type 2      segment   \n",
       "208580      NC_001354  human  Human papillomavirus type 41      segment   \n",
       "208577      NC_001356  human  Human papillomavirus type 1a      segment   \n",
       "194084      NC_001430  human               Enterovirus D68      segment   \n",
       "...               ...    ...                           ...          ...   \n",
       "257436      NC_055340  human                Echarate virus      segment   \n",
       "257437      NC_055341  human                Echarate virus      segment   \n",
       "257464      NC_055342  human               Maldonado virus      segment   \n",
       "257465      NC_055343  human               Maldonado virus      segment   \n",
       "257463      NC_055344  human               Maldonado virus      segment   \n",
       "\n",
       "                  family                genus  \n",
       "810        Herpesviridae       Varicellovirus  \n",
       "214039  Papillomaviridae  Alphapapillomavirus  \n",
       "208580  Papillomaviridae     Nupapillomavirus  \n",
       "208577  Papillomaviridae     Mupapillomavirus  \n",
       "194084    Picornaviridae          Enterovirus  \n",
       "...                  ...                  ...  \n",
       "257436     Phenuiviridae          Phlebovirus  \n",
       "257437     Phenuiviridae          Phlebovirus  \n",
       "257464     Phenuiviridae          Phlebovirus  \n",
       "257465     Phenuiviridae          Phlebovirus  \n",
       "257463     Phenuiviridae          Phlebovirus  \n",
       "\n",
       "[223 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter viral family that are only found in bats\n",
    "viral_fam = batvir_df[\"Viral family\"].dropna().unique().tolist()\n",
    "\n",
    "# filtering main data frame to only viruses found in all bats \n",
    "sel_dfs = []\n",
    "for v_fam in viral_fam:\n",
    "    sel_df = ncbi_acc_df.loc[ncbi_acc_df[\"Selected lineage\"].str.contains(v_fam)]\n",
    "    if len(sel_df) == 0:\n",
    "        continue\n",
    "    sel_dfs.append(sel_df)\n",
    "\n",
    "\n",
    "# selecting viral genomes found in bats that are also found in human\n",
    "sel_dfs = pd.concat(sel_dfs, axis=0)\n",
    "sel_dfs = sel_dfs.loc[sel_dfs[\"Host\"].notnull()]\n",
    "sel_dfs = sel_dfs.loc[sel_dfs[\"Host\"].isin([\"human\"])]\n",
    "\n",
    "# splitting the selected lineage into Family, Genius \n",
    "sel_dfs[[\"family\", \"genus\"]] = sel_dfs[\"Selected lineage\"].apply(split_lineages)\n",
    "sel_dfs[\"Representative\"] = sel_dfs[\"Representative\"].apply(clean_accession)\n",
    "sel_dfs = sel_dfs.drop([\"Selected lineage\"], axis=\"columns\")\n",
    "\n",
    "# removing enties that do not have either a family or a genus entry\n",
    "sel_dfs.loc[(sel_dfs[\"family\"].notnull() & (sel_dfs[\"genus\"].notnull()))]\n",
    "\n",
    "# grouping data frames based on accession and removing duplicates\n",
    "groups = sel_dfs.groupby(\"Representative\")\n",
    "cleaned_dfs = []\n",
    "for name, df in groups:\n",
    "    df = df.drop(\"Neighbor\", axis=\"columns\")\n",
    "    df = df.drop_duplicates(subset=[\"Representative\"])\n",
    "    cleaned_dfs.append(df)\n",
    "\n",
    "# generating bat-human viral genome csv\n",
    "init_db = vsp.init_db_path()\n",
    "save_path = str((Path(init_db) / \"filtered_bat_virus.csv.gz\").absolute())\n",
    "cleaned_dfs = pd.concat(cleaned_dfs, axis=0)\n",
    "cleaned_dfs.to_csv(save_path, compression=\"gzip\")\n",
    "cleaned_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing\n",
    "def _call_entrez_viral_genes(email: str, accession: str, buffer=0.5) -> str:\n",
    "    \"\"\"Submits request to NCBI's genes database via entrez portal\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    email : str\n",
    "        valid email address\n",
    "    accession : str\n",
    "        genome accession number\n",
    "    buffer : int, float\n",
    "        Buffer time added when submitting a request in seconds\n",
    "        Default = 0.5\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        TODO: add return type information\n",
    "    \"\"\"\n",
    "\n",
    "    # call genes data \n",
    "\n",
    "    \n",
    "    # internal parser that returns a tuple of ranges \n",
    "    pass\n",
    "\n",
    "def generate_viral_genome_profile(genome_db: str) -> dict:\n",
    "    \"\"\"Generates a json file that profiles \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genome_db : str\n",
    "        path to genome database\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict\n",
    "        viral genome profiles. Also written in JSON format in the genome database\n",
    "    \"\"\"\n",
    "    # gene_positions = _call_entrez_viral_genes(email=email, accession=acc_id)\n",
    "    pass\n",
    "    \n",
    "\n",
    "def _call_entrez_viral_genome(email: str, accession: str, buffer=0.3) -> str:\n",
    "    \"\"\"Submits request to NCBI viral genome database via entrez portal. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    email : str\n",
    "        valid email address\n",
    "    accession : str\n",
    "        genome accession number\n",
    "    buffer : int, float\n",
    "        Buffer time added when submitting a request in seconds\n",
    "        Default = 0.3\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        Genome fasta \n",
    "    \"\"\"\n",
    "    # creating a reference of the default stdout \n",
    "    old_stdout = sys.stdout\n",
    "\n",
    "    # creating a container storing stdout \n",
    "    fasta_result = StringIO()\n",
    "    sys.stdout = fasta_result\n",
    "    \n",
    "    # calling \n",
    "    c = entrezpy.conduit.Conduit(email)\n",
    "    fetch_influenza = c.new_pipeline()\n",
    "    sid = fetch_influenza.add_search({'db' : 'nucleotide', 'term' : f\"{accession}\", 'rettype':'count', 'datetype' : 'pdat'})\n",
    "    fid = fetch_influenza.add_fetch({'retmax' : 10, 'retmode' : 'text', 'rettype': 'fasta'}, dependency=sid)\n",
    "    c.run(fetch_influenza)\n",
    "    sleep(buffer)\n",
    "\n",
    "    # store the string from the stdout into variable\n",
    "    genome = fasta_result.getvalue()\n",
    "\n",
    "    # now redict back stdout to screen\n",
    "    sys.stdout = old_stdout\n",
    "\n",
    "    return genome\n",
    "\n",
    "\n",
    "\n",
    "def get_viral_genomes(email: str, accessions: Union[str, list], buffer=0.5) -> dict:\n",
    "    \"\"\" Downloads all viral gneomes and generates a viral genome profiles\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    email : str\n",
    "        valid email address require to send requests to the NCBI\n",
    "        database using entrez.\n",
    "    accession : Union[str, list]\n",
    "        string or list of accession numbers \n",
    "    buffer : int, float\n",
    "        Buffer time added after submitting a request in seconds\n",
    "        Default = 0.5\n",
    "\n",
    "    Return\n",
    "    ------\n",
    "    None\n",
    "        Generates a genome database under ./db \n",
    "    \"\"\"\n",
    "    print(\"\\nBuilding genome database ...\")\n",
    "    genome_db = vsp.init_genome_db_path()\n",
    "\n",
    "    if isinstance(accessions, str):\n",
    "        accessions = accessions.split()\n",
    "\n",
    "    print(\"Checking if genome database exists ...\")\n",
    "    if genome_db_exist():\n",
    "        print(\"Genome database already exists. Checking for missing files\")\n",
    "        expected = set(accessions)\n",
    "        gdb_files = {gpath.rsplit(\"/\",1)[-1] for gpath in genome_db_files()}\n",
    "        missing = expected - gdb_files\n",
    "\n",
    "        if len(missing) > 0:\n",
    "            print(\"Warning there are some missing genomes\")\n",
    "            for acc_id in missing:\n",
    "                print(f\"> Requesting {acc_id} genome ... \")\n",
    "                viral_genome = _call_entrez_viral_genome(email=email, accession=acc_id, buffer=buffer)\n",
    "                save_genome(accession=acc_id, contents=viral_genome)\n",
    "\n",
    "            return genome_db\n",
    "\n",
    "        else:\n",
    "            print(\"No missing genomes\")\n",
    "            return genome_db \n",
    "\n",
    "    for acc_id in accessions:\n",
    "        print(f\">Requesting {acc_id} genome ... \")\n",
    "        viral_genome = _call_entrez_viral_genome(email=email, accession=acc_id, buffer=buffer)\n",
    "        save_genome(accession=acc_id, contents=viral_genome)\n",
    "\n",
    "    return genome_db\n",
    "\n",
    "    \n",
    "\n",
    "        \n",
    "\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Building genome database ...\n",
      "Checking if genome database exists ...\n",
      "Genome database already exists. Checking for missing files\n",
      "Warning there are some missing genomes\n",
      "> Requesting NC_001348 genome ... \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/Users/erikserrano/Development/prelim/prelim3/VSeek/db/genome'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accession_list = cleaned_dfs[\"Representative\"].tolist()\n",
    "get_viral_genomes(email = \"erikishere3@gmail.com\", accessions=accession_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'set1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/erikserrano/Development/prelim/prelim3/VSeek/analysis/00_get_data.ipynb Cell 9'\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/erikserrano/Development/prelim/prelim3/VSeek/analysis/00_get_data.ipynb#ch0000019?line=0'>1</a>\u001b[0m set1\u001b[39m.\u001b[39mdifference(set2)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'set1' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "set1.difference(set2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_001583'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set1.difference(set2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_001586\n",
      "NC_001583\n",
      "NC_001587\n"
     ]
    }
   ],
   "source": [
    "for i in set1:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_001352'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"/Users/erikserrano/Development/prelim/prelim3/VSeek/db/genome/NC_001352/NC_001352.fasta\".split(\"/\")[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "1f89458deca3ca18591a84fdf1e703d2c43dc698aaa275754e9258486684cef0"
  },
  "kernelspec": {
   "display_name": "Python 3.9.13 ('cyberbio')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "27f4d7ae5f0f63d457ba4d13d72bdcdc6ef6e3071269bc00f46f461f8d576997"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
